### ISSUE TITLE
Daphne closes connection and return 404

### ISSUE THREAD

**Issue by knv** [django/daphne#114]
We get this in daphne error logs:

```2017-05-13 13:58:08,325 DEBUG    HTTP GET request for daphne.response.EdtdafmMUb!WqrtfODWix
2017-05-13 13:58:08,375 DEBUG    HTTP 404 response started for daphne.response.EdtdafmMUb!WqrtfODWix
2017-05-13 13:58:08,376 DEBUG    HTTP close for daphne.response.EdtdafmMUb!WqrtfODWix
2017-05-13 13:58:08,376 INFO     "127.0.0.1" - - [13/May/2017:13:58:07 +0000] "GET /ws/user/ HTTP/1.1" 404 82 "-" "-"
2017-05-13 13:58:08,377 DEBUG    HTTP response complete for daphne.response.EdtdafmMUb!WqrtfODWix
```

This is our `routing.py`:
```
    ...
    route('websocket.receive', ws_message_user, path=r'^/ws/user/'),
    route('websocket.receive', ws_message_driver, path=r'^/ws/driver/'),
    
    route('websocket.connect', ws_add_user, path=r'^/ws/user/'),
    route('websocket.connect', ws_add_driver, path=r'^/ws/driver/'),
    ...
```

This is supervisor conf:
```
[program:daphne_production]
environment=
        DJANGO_SETTINGS_MODULE="zxc.settings.production",
        TZ_LOG_DIR="/var/log/zxc-logs/production/django/",
        DB_USER="db_user",
        DB_PASSWORD="db_password",
        MEDIA_ROOT="/media/",
        STATIC_ROOT="/staticfiles/",
        REDIS_PORT="6379"
command=/opt/venvs/zxc/bin/daphne -v 2 -b 0.0.0.0 -p 8002 zxc.asgi:channel_layer
directory=/web/
autostart=true
autorestart=true
stopasgroup=true
user=www-data
stderr_logfile=/var/log/zxc/supervisor/daphne/error.log
stdout_logfile=/var/log/zxc/supervisor/daphne/daphne.log
```

We have a 8 core server with 19gb ram.

I added more Daphne instances which get the number of 404 responses from 33% to around 1% but again last day it spiked to 26%. Then again today it came down to around 9%.

My questions are:

1. Why do we get these 404s errors? Did we do something wrong in our routing?
2. Why percentage of 404 error decreased when we add more Daphne instances?

Thanks in advance for you time

**Comment by andrewgodwin** [comment]
Getting a 404 response means that you're not negotiating a WebSocket connection; that would manifest as either a WebSocket error or at worst a 403 (no other code comes out of the WebSocket subsystem).

Given that, I'd suggest you look closely at whatever is in front of your server to make sure it is passing through WebSocket correctly - this includes CDNs as well as static file servers like Nginx. I would suggest connecting directly to Daphne to verify this first.

If you have more issues like this, the place for support is the django-users mailing list.

**Comment by knv** [comment]
Thanks for your response.

I also found many occurrence of these two errors in Daphne error logs.

```
2017-05-15 08:57:04,537 ERROR    HTTP/WS send decode error: Cannot dispatch message on channel u'daphne.response.ZaVIzxcNEN!fPNEeiGFuE' (unknown)
```

```
2017-05-15 08:56:59,440 INFO     failing WebSocket opening handshake ('Access denied')
2017-05-15 08:56:59,440 WARNING  dropping connection to peer tcp4:127.0.0.1:39857 with abort=False: Access denied
2017-05-15 08:56:59,440 DEBUG    WebSocket daphne.response.ZaVIzxcNEN!SeyHyjEjbA rejected by application
2017-05-15 08:56:59,440 INFO     failing WebSocket opening handshake ('Access denied')
2017-05-15 08:56:59,441 WARNING  dropping connection to peer tcp4:127.0.0.1:39856 with abort=False: Access denied
2017-05-15 08:56:59,441 DEBUG    WebSocket daphne.response.ZaVIzxcNEN!gYlfskYhvF rejected by application
2017-05-15 08:56:59,441 CRITICAL Unhandled Error
Traceback (most recent call last):
  File "/opt/venvs/production/local/lib/python2.7/site-packages/daphne/server.py", line 123, in run
    reactor.run(installSignalHandlers=self.signal_handlers)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/base.py", line 1199, in run
    self.mainLoop()
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/base.py", line 1211, in mainLoop
    self.doIteration(t)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/epollreactor.py", line 235, in doPoll
    log.callWithLogger(selectable, _drdw, selectable, fd, event)
--- <exception caught here> ---
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/python/log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/python/log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 610, in _doReadOrWrite
    self._disconnectSelectable(selectable, why, inRead)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 255, in _disconnectSelectable
    selectable.connectionLost(f)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/twisted/internet/tcp.py", line 292, in connectionLost
    protocol.connectionLost(reason)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/autobahn/twisted/websocket.py", line 129, in connectionLost
    self._connectionLost(reason)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/autobahn/websocket/protocol.py", line 2437, in _connectionLost
    WebSocketProtocol._connectionLost(self, reason)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/autobahn/websocket/protocol.py", line 1104, in _connectionLost
    self._onClose(self.wasClean, WebSocketProtocol.CLOSE_STATUS_CODE_ABNORMAL_CLOSE, "connection was closed uncleanly (%s)" % self.wasNotCleanReason)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/autobahn/twisted/websocket.py", line 172, in _onClose
    self.onClose(wasClean, code, reason)
  File "/opt/venvs/production/local/lib/python2.7/site-packages/daphne/ws_protocol.py", line 209, in onClose
    self.cleanup()
  File "/opt/venvs/production/local/lib/python2.7/site-packages/daphne/ws_protocol.py", line 234, in cleanup
    del self.factory.reply_protocols[self.reply_channel]
exceptions.KeyError: u'daphne.response.ZaVIzxcNEN!SeyHyjEjbA'
```

In both cases, it seems to me - and I can be easily wrong - that corresponding web socket connection of the reply channel has been closed and when Daphne tries to send message on that channels this error is raised. 

I don't know if these errors are related to my previous issue with 404 errors. Again I can only guess that when Daphne crashes Nginx can't find the upstream and returns 404.

I will be grateful if you could shed some light on this issue.

**Comment by andrewgodwin** [comment]
That last traceback is caused by a connection being stopped before Daphne even starts the handshake, I have committed a simple fix in eb195c6
